modules/gapi/src/streaming/onevpl/file_data_provider.cpp

```diff
@@ -3,44 +3,152 @@
+
 #include <errno.h>
 
 #include "streaming/onevpl/file_data_provider.hpp"
+#include "streaming/onevpl/cfg_params_parser.hpp"
+#include "streaming/onevpl/utils.hpp"
+#include "logger.hpp"
 
 namespace cv {
 namespace gapi {
 namespace wip {
 namespace onevpl {
 
-FileDataProvider::FileDataProvider(const std::string& file_path) :
-    source_handle(fopen(file_path.c_str(), "rb"), &fclose) {
+#ifdef HAVE_ONEVPL
+FileDataProvider::FileDataProvider(const std::string& file_path,
+                                   const std::vector<CfgParam> codec_params,
+                                   uint32_t bitstream_data_size_value) :
+    source_handle(nullptr, &fclose),
+    bitstream_data_size(bitstream_data_size_value) {
+
+    GAPI_LOG_DEBUG(nullptr, "[" << this << "] " <<
+                            "check codec Id from CfgParam, total param count: " <<
+                            codec_params.size());
+    auto codec_it =
+        std::find_if(codec_params.begin(), codec_params.end(), [] (const CfgParam& value) {
+            return value.get_name() == "mfxImplDescription.mfxDecoderDescription.decoder.CodecID";
+        });
+    if (codec_it == codec_params.end())
+    {
+        GAPI_LOG_WARNING(nullptr, "[" << this << "] " <<
+                                  "\"mfxImplDescription.mfxDecoderDescription.decoder.CodecID\" "
+                                  "is absent, total param count" << codec_params.size());
+        throw DataProviderUnsupportedException("\"mfxImplDescription.mfxDecoderDescription.decoder.CodecID\" "
+                                               "is required for FileDataProvider");
+    }
+
+    codec = cfg_param_to_mfx_variant(*codec_it).Data.U32;
+
+    GAPI_LOG_DEBUG(nullptr, "[" << this << "] " <<
+                            "opening file: " << file_path);
+    source_handle.reset(fopen(file_path.c_str(), "rb"));
     if (!source_handle) {
         throw DataProviderSystemErrorException(errno,
                                                "FileDataProvider: cannot open source file: " + file_path);
     }
+
+    GAPI_LOG_INFO(nullptr, "[" << this << "] " <<
+                            "file: " << file_path << " opened, codec requested: " << mfx_codec_id_to_cstr(codec));
 }
 
 FileDataProvider::~FileDataProvider() = default;
 
-size_t FileDataProvider::fetch_data(size_t out_data_bytes_size, void* out_data) {
+IDataProvider::mfx_codec_id_type FileDataProvider::get_mfx_codec_id() const {
+    return codec;
+}
+
+bool FileDataProvider::fetch_bitstream_data(std::shared_ptr<mfx_bitstream> &out_bitstream) {
+
+    GAPI_LOG_DEBUG(nullptr, "[" << this << "] " <<
+                            ", dst: " << out_bitstream.get());
     if (empty()) {
-        return 0;
+        return false;
+    }
+
+    if (!out_bitstream) {
+        out_bitstream = std::make_shared<mfx_bitstream>();
+        out_bitstream->MaxLength = bitstream_data_size;
+        out_bitstream->Data = (mfxU8 *)calloc(out_bitstream->MaxLength, sizeof(mfxU8));
```

Comment:
if `mxf_bitstream` structure is yours, I'd avoid unnecessary allocations outside of it

If it is not, I'd find a way to associate an automatic buffer with a wrapper over this structure and set `Data` to that buffer pointer.

Reply:
second option:
it allocates at once for whole source, effort is more that profit

Reply:
allocating it once for the whole source changes the overall allocation model, what I ask for above is just making what's now automatic

Reply:
I agree, that it would be quite worth for regular usage, but otherwise it doesn't use any pre-allocation in case of async demux, because it uses mapping from MFP buffers, so byte-receiving constructor would be useless for this case. 
From my perspective, implementing smart-class `mxf_bitstream` for both cases looks redundant at the moment and should be done by demand when it would be used more than in 1-2 places